### PR TITLE
Refactoring of AppsController routes

### PR DESCRIPTION
**Pull request by meln1k**
Summary: Moved routes structure in a different commit, so it will be easier to review separate controller method implementation

JIRA issues: MARATHON-7299

### CODE REVIEW

**Comment by timcharper** [comment]
I have mixed feelings :) On the one hand, I see the value in being able to focus on reviewing and negotiating routes, on the other hand, it seems like this approach compromises the atomicity of our commits?

Just a thought. I don't have an extremely strong objection :) Interested to hear what others on the team think.

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala [hunk 1]
```diff
@@ -50,93 +53,101 @@ class AppsController(
...
-    authorized(CreateRunSpec, app).apply {
-      onSuccess(create) { (plan, app) =>
-        //TODO: post ApiPostEvent
-        complete((StatusCodes.Created, Seq(Headers.`Marathon-Deployment-Id`(plan.id)), app))
-      }
-    }
-  }
+  private def listApps(implicit identity: Identity): Route = ???
 
-  private def showApp(appId: PathId)(implicit identity: Identity): Route = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala]: Why is this removed?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 2]
```diff
@@ -0,0 +1,1433 @@
...
+      val entity = HttpEntity(body).withContentType(ContentTypes.`application/json`)
+
+      Post(Uri./, entity) ~> route ~> check {
+        Then("It is successful")
+        status shouldEqual StatusCodes.Created
+        header[Headers.`Marathon-Deployment-Id`] should not be 'empty
+
+        And("the JSON is as expected, including a newly generated version")
+        import mesosphere.marathon.api.v2.json.Formats._
+        val expected = AppInfo(
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: How about using a JSON string here? What do you think? This avoids serialization

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 3]
```diff
@@ -0,0 +1,1433 @@
...
+          normalizeAndConvert(app).copy(versionInfo = VersionInfo.OnlyVersion(clock.now())),
+          maybeTasks = Some(immutable.Seq.empty),
+          maybeCounts = Some(TaskCounts.zero),
+          maybeDeployments = Some(immutable.Seq(Identifiable(plan.id)))
+        )
+
+        Try(JsonTestHelper.assertThatJsonString(responseAs[String]).correspondsToJsonOf(expected))
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Why wrapping it in a try?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 4]
```diff
@@ -0,0 +1,1433 @@
...
+
+    "Create a new app with IP/CT, no default network name, Alice does not specify a network" in new Fixture {
+      Given("An app and group")
+      val app = App(
+        id = "/app",
+        cmd = Some("cmd"),
+        networks = Seq(Network(mode = NetworkMode.Container))
+      )
+      the[NormalizationException] thrownBy {
+        prepareApp(app, groupManager)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Is this the right spot for testing this? Isn't this independent of the controller?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 5]
```diff
@@ -0,0 +1,1433 @@
...
+      clock += 5.seconds
+      val entity = HttpEntity(body).withContentType(ContentTypes.`application/json`)
+      Post(Uri./, entity) ~> route ~> check {
+        Then("It is successful")
+        status shouldEqual StatusCodes.Created withClue s"body = ${ByteString(body).utf8String}, response = ${responseAs[String]}"
+        header[Headers.`Marathon-Deployment-Id`] should not be 'empty
+
+        And("the JSON is as expected, including a newly generated version")
+        import mesosphere.marathon.api.v2.json.Formats._
+        val expected = AppInfo(
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: How about avoiding the serialization.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 6]
```diff
@@ -0,0 +1,1433 @@
...
+import scala.language.postfixOps
+import scala.util.Try
+import scala.util.control.NonFatal
+
+class AppsControllerTest extends UnitTest with GroupCreation with ScalatestRouteTest {
+
+  case class Fixture(
+      clock: SettableClock = new SettableClock(),
+      auth: TestAuthFixture = new TestAuthFixture {
+        override val UnauthorizedStatus = 403
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Yep, it is! So it is temporary workaround until we fix all the other tests depending on it

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 7]
```diff
@@ -0,0 +1,1433 @@
...
+      healthCheckManager = healthCheckManager,
+      instanceTracker = instanceTracker,
+      taskKiller = taskKiller,
+      config = config,
+      groupManager = groupManager,
+      pluginManager = PluginManager.None
+    )
+
+    implicit val rejectionHandler: RejectionHandler = AkkaHttpMarathonService.rejectionHandler
+    val route: Route = Route.seal(appsController.route)
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: We want to make sure that our own rejection handler is used instead of some default.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 8]
```diff
@@ -0,0 +1,1433 @@
...
+      val migrated = AppNormalization.forDeprecated(normalizationConfig).normalized(app)
+      AppNormalization(normalizationConfig).normalized(migrated)
+    }
+
+    def normalizeAndConvert(app: App): AppDefinition = {
+      val normalized = normalize(app)
+      Raml.fromRaml(normalized)
+    }
+
+    def prepareApp(app: App, groupManager: GroupManager): (Array[Byte], DeploymentPlan) = {
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: If I understand it correctly, it will update mocked group manager with appropriate values for the given app.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 9]
```diff
@@ -0,0 +1,1433 @@
...
+           |  }
+           |}
+      """.stripMargin
+
+      When("The request is processed")
+      HttpEntity(body).withContentType(ContentTypes.`application/json`)
+    }
+  }
+
+  case class FixtureWithRealGroupManager(
```
[the remark this answers is not in the document]
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Because we need a real implementation here to simulate a non-existing app.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 2 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: I don't know what's wrong with serialization, but yes, I can do it.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 5 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: I'm not sure it's possible because deployment id is generated dynamically.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 2 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: It seems it's not possible because of the dynamic id generation.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 5 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: The can be easily set with string interpolation.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Aren't we breaking the API then?

[the remark this answers is not in the document]
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Also, I've just checked. This value is never used. I'd then remove these changes,

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Hm, this seems too much for a test to me but I understand that you've just copied it. Since I cannot think of a better solution for now I'm fine with it.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala (hunk 2 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: Just used string interpolation :).

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala [hunk 10]
```diff
@@ -0,0 +1,1856 @@
...
+          |        "provider": "dvdi",
+          |        "name": "namedfoo",
+          |        "options": {"dvdi/driver": "bar"}
+          |      },
+          |      "mode": "RW"
+          |    }]
+        """.stripMargin, groupManager, auth
+      )
+
+      Then("The return code indicates create success")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: I think we should check the rejections instead. Let's decide tomorrow to follow up. I don't see this as a blocker for this diff.

Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/AppsControllerTest.scala]: I think these should check the rejections.
